src/nvim/memory.h

```diff
@@ -2,6 +2,107 @@
 #define NVIM_MEMORY_H
 
 #include <stddef.h>  // for size_t
+#include <string.h>
+#include <strings.h>
+
+#include "nvim/func_attr.h"
+
+#ifdef HAVE_CONFIG_H
+# include "auto/config.h"  // for the HAVE_* macros
+#endif
+
+// defines to avoid typecasts from (char_u *) to (char *) and back
+// (vim_strchr() and vim_strrchr() are now in alloc.c)
+#define STRLEN(s)           strlen((char *)(s))
+#define STRCPY(d, s)        strcpy((char *)(d), (char *)(s))
+#define STRNCPY(d, s, n)    strncpy((char *)(d), (char *)(s), (size_t)(n))
+#define STRLCPY(d, s, n)    xstrlcpy((char *)(d), (char *)(s), (size_t)(n))
+#define STRCMP(d, s)        strcmp((char *)(d), (char *)(s))
+#define STRNCMP(d, s, n)    strncmp((char *)(d), (char *)(s), (size_t)(n))
+
+#if defined(HAVE_STRCASECMP)
+# define STRICMP(d, s)      strcasecmp((char *)(d), (char *)(s))
+#elif defined(HAVE_STRICMP)
+# define STRICMP(d, s)     stricmp((char *)(d), (char *)(s))
+#elif defined(HAVE_U_STRICMP)
+# define STRICMP(d, s)     _stricmp((char *)(d), (char *)(s))
+#else
+# error "strcasecmp/stricmp are undefined. Please file a bug report."
+#endif
+
+/// Like strcpy() but allows overlapped source and destination.
+#define STRMOVE(d, s)       memmove((d), (s), STRLEN(s) + 1)
+
+#if defined(HAVE_STRNCASECMP)
+# define STRNICMP(d, s, n)  strncasecmp((char *)(d), (char *)(s), (size_t)(n))
+#elif defined(HAVE_STRNICMP)
+# define STRNICMP(d, s, n) strnicmp((char *)(d), (char *)(s), (size_t)(n))
+#elif defined(HAVE_U_STRNICMP)
+# define STRICMP(d, s)     _strnicmp((char *)(d), (char *)(s))
+#else
+# error "strncasecmp/strnicmp are undefined. Please file a bug report."
+#endif
+
+#define STRCAT(d, s)        strcat((char *)(d), (char *)(s))
+#define STRNCAT(d, s, n)    strncat((char *)(d), (char *)(s), (size_t)(n))
+
+#define vim_strpbrk(s, cs) (char_u *)strpbrk((char *)(s), (char *)(cs))
+
+// FIXME(SplinterOfChaos): the strdup functions should be used without the x-
+// prefix, but even when the feature test passes, they may cause errors.
```

Comment:
At first, I wanted to just make these functions `strdup()`, `stpcpy()`, etc., to make our code look more like standard C. I can't remember exactly what error occurred, but I thought it'd be less confusing to people on platforms that don't have `strdup()` if we kept the `x` prefix, anyway. I removed this comment.